File: Sources/windows/BitTherapyCLR/src/game/entity.cpp
#include "entity.h"

#include <chrono>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include "../game/game.h"
#include "../sprites/sprites.h"
#include "../utils/utils.h"

Entity::Entity(
    uint32_t id,
    double fps, 
    double settingsSpeedMultiplier,
    const Species* species, 
    const SpriteSet* spriteSet, 
    Rect frame
) :
    id(id),
    fps(fps),
    speed(0.0),
    species(species), 
    spriteSet(spriteSet), 
    frame(frame),
    direction(Vector2d(1.0, 0.0)),
    capabilities(std::vector<std::shared_ptr<EntityCapability>>()),
    currentSprite(Sprite("", std::vector<std::string>({}), 1.0))
{
    auto movementSprite = species->movementPath;
    setupSpeed(settingsSpeedMultiplier);
    changeSprite(movementSprite); 
}

void Entity::addCapability(std::shared_ptr<EntityCapability> capability) {
    capabilities.push_back(capability);
}

const std::string Entity::currentSpriteFrame() const {
    return currentSprite.currentFrame();
}

void Entity::update(std::chrono::milliseconds timeSinceLastUpdate) {
    for (auto& capability : capabilities) {
        capability->update(timeSinceLastUpdate, this);
    }
    currentSprite.update(timeSinceLastUpdate);
}

uint32_t Entity::changeSprite(const std::string& animationName) {
    if (currentSprite.animationName != animationName) {
        currentSprite = spriteSet->sprite(animationName, fps);
    }
    return 1000 * currentSprite.numberOfFrames() / fps;
}

std::string Entity::description() const {
    auto spriteName = fileName(currentSpriteFrame());
    std::stringstream ss; 

    ss << "Entity @ " << this << std::endl;
    ss << "  Species: " << species->id << std::endl;
    ss << "  Sprite: " << spriteName << std::endl;
    ss << "  dx: " << direction.x << std::endl;
    ss << "  dy: " << direction.y << std::endl;
    ss << "  x: " << frame.x << std::endl;
    ss << "  y: " << frame.y << std::endl;
    ss << "  w: " << frame.w << std::endl;
    ss << "  h: " << frame.h << std::endl;
    return ss.str();
}

void Entity::setupSpeed(double settingsSpeedMultiplier) {
    speed = 30.0 * species->speed * settingsSpeedMultiplier;
}

File: Sources/windows/App/Sources/Game/Entity.cs
using App;
using App;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Policy;
using System.Text;
using System.Threading.Tasks;

namespace App
{
    public class Entity
    {
        private double fps;
        private SpriteSet spriteSet;
        private List<EntityCapability> capabilities;
        private Sprite currentSprite;

        private void SetupSpeed(double speedMultiplier)
        {
            Speed = 30.0 * Species.Speed * speedMultiplier;
        }

        public uint Id { get; set; }
        public Rect Frame { get; set; }
        public Vector2d Direction { get; set; }
        public double Speed { get; set; }
        public Species Species { get; set; }

        public Entity(uint id, double fps, double speedMultiplier, Species species, SpriteSet spriteSet, Rect frame)
        {
            Id = id;
            this.fps = fps;
            Speed = 0.0;
            Species = species;
            this.spriteSet = spriteSet;
            Frame = frame;
            Direction = new Vector2d(1.0, 0.0);
            capabilities = new List<EntityCapability>();
            currentSprite = new Sprite("", new List<string>(), 1.0);

            var movementSprite = species.MovementPath;
            SetupSpeed(speedMultiplier);
            ChangeSprite(movementSprite);
        }

        public string CurrentSpriteFrame()
        {
            return currentSprite.CurrentFrame;
        }

        public void Update(TimeSpan timeSinceLastUpdate)
        {
            foreach (var capability in capabilities)
            {
                capability.Update(timeSinceLastUpdate, this);
            }
            currentSprite.Update(timeSinceLastUpdate);
        }

        public void AddCapability(EntityCapability capability)
        {
            capabilities.Add(capability);
        }

        public uint ChangeSprite(string animationName)
        {
            if (currentSprite.AnimationName != animationName)
            {
                currentSprite = spriteSet.CreateSprite(animationName, fps);
            }
            return (uint)(1000 * currentSprite.NumberOfFrames / fps);
        }

        public string Description()
        {
            var spriteName = FileUtils.FileName(CurrentSpriteFrame());
            var sb = new StringBuilder();

            sb.AppendLine($"Entity @ {this}");
            sb.AppendLine($"  Species: {Species.Id}");
            sb.AppendLine($"  Sprite: {spriteName}");
            sb.AppendLine($"  dx: {Direction.X}");
            sb.AppendLine($"  dy: {Direction.Y}");
            sb.AppendLine($"  x: {Frame.X}");
            sb.AppendLine($"  y: {Frame.Y}");
            sb.AppendLine($"  w: {Frame.W}");
            sb.AppendLine($"  h: {Frame.H}");

            return sb.ToString();
        }
    }
}
